File: DAC/minMax.cpp
#include <bits/stdc++.h>
using namespace std;

// Structure to store min and max
struct MinMax {
    int min;
    int max;
};

// Function to find min and max using Divide and Conquer
MinMax findMinMax(vector<int>& arr, int low, int high) {
    MinMax result, left, right;

    // Base case: Only one element
    if (low == high) {
        result.min = result.max = arr[low];
        return result;
    }

    // Base case: Two elements
    if (high == low + 1) {
        result.min = min(arr[low], arr[high]);
        result.max = max(arr[low], arr[high]);
        return result;
    }

    // Divide: Find the middle index
    int mid = low + (high - low) / 2;

    // Conquer: Recursively find min and max in left and right halves
    left = findMinMax(arr, low, mid);
    right = findMinMax(arr, mid + 1, high);

    // Combine: Get final min and max
    result.min = min(left.min, right.min);
    result.max = max(left.max, right.max);

    return result;
}

int main() {
    vector<int> arr = {3, 5, 1, 8, 9, 2, 7};
    int n = arr.size();

    MinMax result = findMinMax(arr, 0, n - 1);

    cout << "Minimum: " << result.min << endl;
    cout << "Maximum: " << result.max << endl;

    return 0;
}

File: Model_2/DAC/MM/MinMax.java
package Model_2.DAC.MM;


class MinMax {
    int Min;
    int Max;

    MinMax(int min, int max) {
        this.Min = min;
        this.Max = max;
    }

    static MinMax solveMinMax(int a[], int low, int high) {
            if ( high - low == 1 ) {
                if ( a[low] > a[high] ) return new MinMax(a[high], a[low]);
                else return new MinMax(a[low], a[high]);
            } 
            if ( high - low == 0 ) {
                return new MinMax(a[low],a[low]);
            } 
            int mid = (low + high) / 2;
            MinMax left = solveMinMax(a, low, mid) ;
            MinMax right = solveMinMax(a, mid + 1, high);

            int realMin, realMax;
            if(left.Max > right.Max) realMax = left.Max;
            else realMax = right.Max;
            if ( left.Min < right.Min ) realMin = left.Min;
            else realMin = right.Min;

            return new MinMax(realMin, realMax);
    } 

    public static void main(String[] args) {
        int a[] = {100,5,2,7,3,1,-3,5,30};
        System.out.println(solveMinMax(a, 0, a.length - 1).Min + " " + solveMinMax(a, 0, a.length - 1).Max);
    }
}

